[#60183] add pattern contract validation to subject configuration

Kharonus opened the pull request:
# Ticket
[OP#60183](https://community.openproject.org/work_packages/60183)

# What are you trying to accomplish?
- subject configuration form for work package type must react on data contract validation
- allow persistence of pattern blueprint if configuration is set to `manual`

# What approach did you choose and why?
- added data contract validation to service
- use specific controller & action to update form configuration
- introduce form object model
- harmonized route definition

Kharonus committed:
[#60183] add pattern contract validation to subject configuration

- https://community.openproject.org/work_packages/60183
- rewrite work package type routes
- added new controller for subject configuration
- fixed test expectation values for radio buttons
- removed tests for disabled submit button
- extracted tests to new controller

NobodysNightmare commented on app/forms/work_packages/types/subject_configuration_form.rb:
```diff
@@ -31,39 +31,33 @@
             label: I18n.t("types.edit.subject_configuration.automatically_generated_subjects.label"),
             caption: I18n.t("types.edit.subject_configuration.automatically_generated_subjects.caption"),
-            disabled: !EnterpriseToken.active? && !has_pattern?,
+            disabled: !EnterpriseToken.active?,
```
:yellow_circle: I am pretty sure that the `&& !has_pattern` was required to fulfill our requirements around OP#59929. Specifically: If the token is removed, but the pattern is enabled, we should keep editing enabled.

Kharonus replied:
Yes, I removed it, because I introduced the form object and no longer worked on top of a wp type model.

I added now the check again in a slightly changed way, but expressing a similar logic for disabling the radio button.

NobodysNightmare commented on app/helpers/tabs_helper.rb:
```diff
@@ -37,8 +39,8 @@ def render_tabs(tabs, form = nil, with_tab_nav: true)
-  def selected_tab(tabs)
-    tabs.detect { |t| t[:name] == params[:tab] } || tabs.first
+  def selected_tab(tabs, default_tab = nil)
+    tabs.detect { |t| t[:name] == params[:tab] || t[:name] == default_tab } || tabs.first
```
:yellow_circle: I didn't yet see where this is called, but won't this cause the default tab to be returned, even though there might be a tab that would fit the `params[:tab]`?

E.g.

```ruby
tabs = %w[a b c d].map { |name| { name: name } }
params = { tab: "c" }
selected_tab(tabs, "b")
=> "b"
```

I would expect that you want the above example to return "c", but the implementation will return "b".

Kharonus replied:
true.... why is that happening? I will look into it. Thanks for finding it.

NobodysNightmare replied:
Explanation-wise: Because both conditions are in a single `detect` (aka `find`) block. I guess what you intended to do was `tabs.find { /* by params */ } || tabs.find { /* by default */ } || tabs.first`

Kharonus replied:
solved

NobodysNightmare commented on app/services/base_type_service.rb:
```diff
@@ -75,13 +75,17 @@ def set_params_and_validate(params)
+
+    # Set patterns includes a data validation before assigning the value to the attribute.
+    # A failure should return a failure.
+    set_patterns(params) unless params[:patterns].nil?
```
:green_circle: I'd prefer to avoid negations. Therefore I'd suggest either of

```suggestion
    set_patterns(params) if params[:patterns]
```

or

```suggestion
    set_patterns(params) if params[:patterns].present?
```

None of the two is exactly equal to your code w.r.t. semantics, thought I'd assume that we either expect a hash with some keys or `nil`, in which case the behaviour should be the same.

Kharonus replied:
yep, agreed

NobodysNightmare commented on app/controllers/work_packages/types/subject_configuration_controller.rb:
```diff
@@ -0,0 +1,83 @@
+                            .to_h
+
+        UpdateTypeService.new(@type, current_user)
+                         .call({ patterns: pattern_collection_update(form_params) }) do |call|
```
:red_circle: I think this code is not forward-compatible enough for the data structure we prepared. Let me explain...

Context: The `patterns` attribute was written in a way, that it already expects us to handle more than just the patterns of a subject (the focus of our current story). You can see that, because it's not the `subject_pattern` attribute, but rather it's an attribute that looks like a hash and has a key called `subject`. This means we can expect that the `patterns` attribute is supposed to be used for other kinds of patterns in the future and `subject` is just one of them.

Now this code is calling the `UpdateTypeService` with a full representation of `patterns`. The `UpdateTypeService` will take this to overwrite the `patterns` attribute in full. On the other hand, we source the value from `pattern_collection_update`, which only takes the `form_params` into account, which will only carry subject-specific state here (our controller is the **subject** configuration controller).

In other words: This controller is written in a way that it will overwrite any and all non-subject-patterns, if they existed on the `patterns` collection. IMO that's an error.

The two high-level solutions that come to my mind are:
1. Write this update in a way that non-subject keys will not be harmed
2. Change the structure of `patterns` back to something that does not claim to be future-proofed for other things than subjects

My preferred solution would be 1, but I remember that @mereghost talked to me about not future-proofing too much, that's why I am suggesting 2 as well. The current solution is IMO "half future-proof", which is the worst of both worlds. We now have a data structure that can hold more than just a subject, but we treat it like a data structure that **only ever** holds a subject.

Kharonus replied:
hmm, let me think a step ahead of option 1:

would you prefer adding this logic to the controller or to the service? The controller could merge the `patterns` data hash before sending it to the service which then does the data validation and assigns it to the object before persitence.

On the other hand, one could argue, the controller should not care about such trifles. in the end it is now a specific controller for subject generation configuration. So, the update type service could be the one merging the data. 

For the second approach I fear, the cut may seem clean, but it is not, when it comes to pattern deletion. So, if the update is a removal of a certain pattern. The first approach can be easily written, as the controller knows everything about the context and purpose.

WDYT? @NobodysNightmare @mereghost

NobodysNightmare replied:
As we discussed IRL: What sounds like a plausible approach is for the controller to take the `patterns` and inside the patterns only assign to the `:subject` key and then pass the whole thing along to the service.

This means the service does not have to guess the intention of its caller and the controller is not _really_ concerned with state-management too much. It just carefully changes the bits it's interested in.

Kharonus replied:
Tried to implement this approach.

NobodysNightmare commented on app/services/base_type_service.rb:
```diff
@@ -107,12 +120,6 @@ def set_attribute_groups(params)
   end
 
-  def set_subject_pattern(params)
-    type.patterns = type.patterns.update_pattern(:subject,
```
:yellow_circle: Assuming we remain at the current implementation, the `update_pattern` method is now unused and can be removed entirely.

NobodysNightmare reviewed (changes requested):
I left some comments, most of them minor, though one (marked with :red_circle:) I deem important to properly discuss. Looks good to me in general.

One thing I was wondering about: I would have expected this PR to contain further validations on the blueprint (e.g. validating the tokens inside for existence). Was this planned for a separate pull request?

NobodysNightmare commented on app/forms/work_packages/types/subject_configuration_form.rb:
```diff
@@ -31,60 +31,55 @@
         subject_form.submit(
           name: :submit,
           label: I18n.t(:button_save),
-          disabled: !EnterpriseToken.active? && !has_pattern?,
```
This was part of a fix that I specifically performed after QA noticed that the submit button remained enabled, even if the rest of the form is disabled.

I agreed that it's confusing to be able to submit a form that allows no changes to be made.

PR introducing this: https://github.com/opf/openproject/pull/17610

Kharonus replied:
I talked it through with dominic just today. The application of the form model makes the disabling of the submit button obsolete. I can explain, if need be :D

NobodysNightmare replied:
As discussed IRL: I can kinda live with most outcomes, but it would make sense to align our wireframes to match what we want to see, so that QA also knows what to expect.

Kharonus replied:
Yes, I will talk to Dominic and cecile.

Kharonus replied:
As discussed IRL: We leave submit button active and I update the AC.

NobodysNightmare reviewed (changes requested):
`permitted_params.rb` still contains my previous code that allows `subject_pattern` and `subject_configuration` on type.

You can clean this up.

Kharonus committed:
[#60183] merge pattern collection on update subject pattern

- only replace `subject` key in pattern collection from within subject
  configuration controller
- fix tab selection so that default tab does not overwrite parameter
  tab due to order

NobodysNightmare commented on spec/controllers/work_packages/types/subject_configuration_controller_spec.rb:
```diff
@@ -0,0 +1,117 @@
+  context "when the user is not logged in" do
+    let(:user) { User.anonymous }
+
+    it "responds with forbidden" do
```
:green_circle: That's not technically what we expect.

The expectation is about a redirect to the login, not an HTTP 403 Forbidden.

Maybe:
```suggestion
    it "requires login" do
```

NobodysNightmare commented on spec/controllers/work_packages/types/subject_configuration_controller_spec.rb:
```diff
@@ -0,0 +1,117 @@
+    end
+  end
+
+  describe "PUT #update_subject_configuration" do
```
:yellow_circle: Both contexts (when the user is not an admin and when the user is not logged in) above are testing the behaviour of the `update_subject_configuration` action when receiving a PUT. I think those contexts should semantically be placed inside this `describe` as well.

As far as I can tell, this should also not conflict with the additional setup peformed in the current describe block.

Kharonus replied:
While this is true, that it probably won't clash, the two contexts above are about controller before_actions. So, technically, those are not about this PUT, we just need any action to trigger the before action. Would you be fine to leave them outside?

NobodysNightmare replied:
Let me add this perspective: That these things are `before_action`s is an implementation detail. A `before_filter` can take an `only` or `except` argument, there is no guarantee that it runs for each action.

So I'd probably ask a proper unit test for a controller to repeat this expectation for all of the controller actions, to ensure that all actions behave the way they should. That's also the reason why I saw this context inside the action `describe`.

I am definitely not going to die on that hill, i.e. I'd be fine to leave the context where it is (especially considering that we only have one action so far).

NobodysNightmare commented on app/services/update_type_service.rb:
```diff
@@ -35,4 +37,24 @@ def call(params)
+      .build(patterns: params[:patterns])
+      .either(
+        ->(collection) { type.patterns = collection },
+        ->(result) { type.errors.add(:patterns, result.errors(full: true).messages.join(", ")) }
```
:green_circle: Sorry that I didn't see this one before. Shouldn't we loop and add all the errors, so that formatting (i.e. the `join(', ')`) is left as a choice to the final renderer?

I.e.
```suggestion
        ->(result) { result.errors(full: true).each { |e| type.errors.add(:patterns, e.message) }
```

Kharonus replied:
Solved in #17644

NobodysNightmare reviewed (approved):
As I said, none of my remaining comments is blocking a merge. LGTM.
